README
#$Id: README 1214 2008-02-09 23:41:58Z jimk $
Perl6::Say - 'print' -- but no newline needed

This document refers to version 0.13 of Perl6::Say.  
This version was released March 16, 2009.

To install this module on your system, place the tarball archive file in a 
temporary directory and call the following:

% gunzip Perl6-Say-0.13.tar.gz
% tar xf Perl6-Say-0.13.tar
% cd Perl6-Say-0.13
% perl Makefile.PL
% make
% make test
% make install

If during installation you wish to view more information on test results, 
substitute the following for the sixth line in the sequence of commands
above:

% make test TEST=VERBOSE

If you are installing this module over any earlier version, you may 
substitute the following for the last line in the sequence of commands 
above:

% make install UNINST=1

If you are installing this module on a Win32 system with 'nmake', 
substitute 'nmake' for 'make' in the sequence of commands above.

In sending e-mail to the maintainer, please put "Perl6::Say", 
"Perl6-Say" or "Say" in the subject line.

Original author:  Damian Conway

Maintainer since July 2006 (version 0.05_002):  James E Keenan
CPAN ID:  JKEENAN
jkeenan@cpan.org

Dependencies:  Uses only modules found in the Perl Core as of 5.007003.
For use with older Perls, see Makefile.PL.

Copyright (C) 2004 Damian Conway.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 


Changes
#$Id: Changes 1344 2009-03-16 03:27:56Z jimk $
Revision history for Perl extension Perl6::Say.

0.01  Thu Mar 25 22:00:02 2004
	- original version; created by h2xs 1.22 with options
		-A -P -X -f -n Perl6::Say

0.02	Sun Mar 28 13:11:42 2004
	- Fixed bug in Makefile (thanks Stefan!)

0.03	Thu Mar 10 23:26:14 2005
    - Changed $\ interaction in line with changes in Perl 6 design
      (C<say @stuff> is now always equivalent to C<print @stuff, "\n">)
    - Fixed multiple export (added import()) (thanks Schwern)

0.04	Fri Apr  8 04:03:10 2005
    - Added correct handling of default call to use $_, like print
      (thanks Schwern)

0.05	Mon Apr 25 22:42:54 2005


0.5.1  Sun Jul 16 08:06:28 2006
    - Fixed doc nit
    - Corrected handling of C<select>'d default filehandles
    - Fixed test using Test::Builder (thanks David, Jim, et al.)

0.06 Jul 27 2006 (I think)
    - First non-beta upload to CPAN by co-maintainer James E Keenan
    - Many tests added and documentation expanded
    - Now properly handles select-ed filehandles

0.07 Wed Sep  6 21:49:18 EDT 2006
    - Made Makefile.PL non-executable, per CPANTS recommendation.

0.08 Sat May  5 12:51:29 EDT 2007
    - Added SKIP blocks to 3 test files (05-, 07-, 09-) to skip tests that
      require printing to in-memory file.  Modified POD accordingly.

0.09 Sun May  6 21:03:30 EDT 2007
    - Corrected calculation of number of skipped tests in t/09.

0.10 Tue Oct 30 21:35:40 EDT 2007
    - Require 5.6.2.

0.11 Thu Nov  1 19:40:47 EDT 2007
    - Require 5.6.2 in Makefile.PL.

0.12 Sat Feb  9 19:35:26 AST 2008
    - Provide documentation and examples on say() in Perl 5.10.

0.13 Sun Mar 15 22:45:45 EDT 2009
    - Regenerate to get latest META.yml format (per bug report from D.A.
      Patrick:  https://rt.cpan.org/Public/Bug/Display.html?id=44102).

0.14 Sun Mar 15 23:25:13 EDT 2009
    - Re-did change in 0.13.

0.15 Sun Mar 29 21:23:44 EDT 2009
    - Require 5.8, mainly to shut up CPAN test that is failing on 5.6.2.

0.16 Fri Apr 30 22:13:09 PDT 2010
    - Removing this distribution fron CPAN is not a good idea since there are applications
      and modules depending on this module exisiting on CPAN. (Tatsuhiko Miyagawa)
      See the thread http://www.nntp.perl.org/group/perl.module-authors/2010/04/msg8542.html


t/lib/Perl6/Say/Auxiliary.pm
package Perl6::Say::Auxiliary;
#$Id: Auxiliary.pm 1213 2008-02-09 23:40:34Z jimk $
# Contains test subroutines for distribution with Perl6::Say
# As of:  March 16, 2009
use strict;
require Exporter;
our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(
    _validate
    capture_say
    capture_say_file
    capture_say_scalar
);
use Carp;
*is = *Test::More::is;

sub _validate {
    my $pred = shift;
    croak "2nd argument to capture_STDOUT must be integer"
        unless $pred =~ m/^\d+$/;
    return ($pred == 1) ? q{line} : q{lines};
}

sub capture_say {
    my $argsref = shift;
    my ($str, @list);
    if (ref($argsref->{data}) eq q{ARRAY}) {
        @list = @{$argsref->{data}};
    } else {
        $str  = $argsref->{data};
    }
    my $pred  = $argsref->{pred};
    my $l = _validate($pred);
    my $capture = IO::Capture::Stdout->new();
    $capture->start;
    &{$argsref->{eval}};
    $capture->stop;
    my $cat = join q{}, $capture->read();
    unless ($str) {
        $str = join q{}, @list;
    }
    is($cat, "$str\n", "$pred $l $argsref->{msg}");
}

sub capture_say_file {
    my $argsref = shift;
    my $pred    = $argsref->{pred};
    my $l       = _validate($pred);
    {
        my $tmp = File::Temp->new( DIR => "./t" );
        my $tmpfile = $tmp->filename;
        &{$argsref->{eval}}($tmpfile, $argsref->{data});
        my $slurp;
        open READ, $tmpfile or croak "Cannot open tempfile for reading";
        {
            local $/ = undef;
            $slurp = <READ>;
        }
        close READ or croak "Cannot close tempfile after reading"; 
        my $str;
        (ref($argsref->{data}) eq q{ARRAY})
            ? $str = join q{}, @{$argsref->{data}}
            : $str = $argsref->{data};
        is($slurp, "$str\n", "$pred $l $argsref->{msg}");
    }
}

sub capture_say_scalar {
    my $argsref = shift;
    my $pred    = $argsref->{pred};
    my $l       = _validate($pred);
    my $got = &{$argsref->{eval}}($argsref->{data});
    my $str;
    (ref($argsref->{data}) eq q{ARRAY})
        ? $str = join q{}, @{$argsref->{data}}
        : $str = $argsref->{data};
    is($got, "$str\n", "$pred $l $argsref->{msg}");
}

1;


lib/Perl6/Say.pm
package Perl6::Say;
#$Id: 01-feature.pl 1207 2008-02-09 23:22:29Z jimk $
use strict;
use warnings;
require 5.006_002;
our $VERSION = '0.16';
use IO::Handle;
use Scalar::Util 'openhandle';
use Carp;

sub say {
    my $currfh = select();
    my $handle;
    {
        no strict 'refs';
        $handle = openhandle($_[0]) ? shift : \*$currfh;
        use strict 'refs';
    }
    @_ = $_ unless @_;
    my $warning;
    local $SIG{__WARN__} = sub { $warning = join q{}, @_ };
    my $res = print {$handle} @_, "\n";
    return $res if $res;
    $warning =~ s/[ ]at[ ].*//xms;
    croak $warning;
}

# Handle direct calls...

no strict 'refs';
sub import { *{caller() . '::say'} = \&say; }
use strict 'refs';

# Handle OO calls:

*IO::Handle::say = \&say if ! defined &IO::Handle::say;

1;

#################### DOCUMENTATION #################### 

=head1 NAME

Perl6::Say - C<print()>, but no newline needed

=head1 WARNING

This module is B<deprecated>. If you're using perl 5.10 or later, you can say:

  use feature 'say';

to use the built-in C<say> function. See L<perlfunc/say> for details.

=head1 SYNOPSIS

    # Perl 5 code...

    use Perl6::Say;

    say 'boo';             # same as:  print 'boo', "\n"

    say STDERR 'boo';      # same as:  print STDERR 'boo', "\n"

    STDERR->say('boo');    # same as:  print STDERR 'boo', \n"

    $fh->say('boo');       # same as:  print $fh 'boo', "\n";

    say();                 # same as:  print "$_\n";

    say undef;             # same as:  print "\n";

=head1 DESCRIPTION

=head2 Note for Users of Perl 5.10

You don't need this module.  The Perl 6 C<say> function is available in Perl
5.10 by saying C<use feature 'say';>.  Hence, this module is of interest only
to users of Perl 5.6 and 5.8.

If you have Perl 5.10 installed, see the F<510/> directory in this
distribution for some elementary examples of C<say> taken from C<perldoc
feature>.

=head2 General

Implements a close simulation of the C<say> function in Perl 6,
which acts like C<print> but automatically appends a newline.

Use it just like C<print> (except that it only supports the indirect object
syntax when the stream is a bareword). That is, assuming the relevant
filehandles are open for output, you can use any of these:

    say @data;
    say FH @data;
    FH->say(@data);
    *FH->say(@data);
    (\*FH)->say(@data);
    say $fh, @data;
    $fh->say(@data);

but not any of these:

    say {FH} @data;
    say {*FH} @data;
    say {\*FH} @data;
    say $fh @data;
    say {$fh} @data;

=head2 Additional Permitted Usages

As demonstrated in the test suite accompanying this distribution,
C<Perl6::Say::say()> can be used in all the following situations.

    $string = q{};
    open FH, ">", \$string;
    say FH qq{Hello World};            # print to a string
    close FH;                          # requires Perl 5.8.0 or later

    use FileHandle;
    $fh = FileHandle->new($file, 'w');
    if (defined $fh) {
        say $fh, qq{Hello World};
        $fh->close;
    }

    use IO::File;
    $fh = IO::File->new($file, 'w');
    if (defined $fh) {
        say $fh, qq{Hello World};
        $fh->close;
    }

    $string = q{};
    open FH, ">", \$string;             # requires Perl 5.8.0 or later
    select(FH);
    say qq{Hello World};
    close FH;

=head2 Interaction with Output Record Separator

In Perl 6, S<C<say @stuff>> is exactly equivalent to
S<C<Core::print @stuff, "\n">>.

That means that a call to C<say> appends any output record separator (ORS)
I<after> the added newline (though in Perl 6, the ORS is an attribute of
the filehandle being used, rather than a global C<$/> variable).

=head2 C<IO::Handle::say()>

IO::Handle version 1.27 or later (which, confusingly, is
found in IO distribution 1.23 and later) also implements a C<say>
method.   Perl6::Say provides its own C<say> method to IO::Handle
if C<IO::Handle::say> is not available.

=head2 Usage with Older Perls

As noted above, some aspects of C<Perl6::Say::say()> will not work with
versions of Perl earlier than 5.8.0.  This is not due to any problem with this
module; it is simply that Perl did not support printing to an in-memory file
(C<print \$string, "\n";>) prior to that point.  (Thanks to a CPAN testers
report from David Cantrell for identifying this limitation.)

=head1 WARNING

The syntax and semantics of Perl 6 is still being finalized
and consequently is at any time subject to change. That means the
same caveat applies to this module.

=head1 DEPENDENCIES

No dependencies other than on modules included with the Perl core as of
version 5.8.0.

Some of the files in the test suite accompanying this distribution use
non-core CPAN module IO::Capture::Stdout.  Tests calling IO::Capture::Stdout
methods are enclosed in C<SKIP> blocks and so should pose no obstacle to
installation of the distribution on systems lacking IO::Capture.  (However,
the maintainer strongly recommends IO::Capture for developers who write a lot
of test code.  So please consider installing it!)

=head1 AUTHOR and MAINTAINER

=head2 AUTHOR

Damian Conway (damian@conway.org).

=head2 MAINTAINER

James E Keenan (jkeenan@cpan.org) (effective v0.06, July 2006).

=head1 ACKNOWLEDGMENTS

Thanks to Damian Conway for dreaming this up.  Thanks to David A Golden for a
close review of the documentation.  Thanks to CPAN tester Jost Krieger for
reporting an error in my SKIP block count in one test file.

=head1 BUGS AND IRRITATIONS

As far as we can determine, Perl 5 doesn't allow us to create a subroutine
that truly acts like C<print>. That is, one that can simultaneously be
used like so:

    say @data;

and like so:

    say {$fh} @data;

Comments, suggestions, and patches welcome.

=head1 COPYRIGHT

Copyright (c) 2004, Damian Conway. All Rights Reserved.
This module is free software. It may be used, redistributed
and/or modified under the same terms as Perl itself.

=cut
